### ISSUE TITLE
lvalue sub flag set when impossible

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9849]
Migrated from [rt.perl.org#68758](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=68758) (status was 'resolved')

Searchable as RT68758$

**Comment by p5pRT** [comment]
### From zefram@fysh.org

#### Created by zefram@fysh.org



Here's a simple case of an lvalue sub&#8203;:

$ perl \-lwe 'my $x; sub foo :lvalue \{ $x \} $x=1; print foo; foo=123; print $x'  
1
123

Here's what happens if I try to add lvalueness to the sub after defining
the body&#8203;:

$ perl \-lwe 'my $x; sub foo \{ $x \} sub foo :lvalue; $x=1; print foo; foo=123; print $x'
1
1

No error\, or even warning\, about using the sub call as an lvalue\, but
it doesn't actually operate as an lvalue\.  Since the lvalue attribute
requires modifying the op tree\, it would be difficult to apply to an
already\-defined sub\, so to attempt it should probably be an error\.

For comparison\, here's what happens if I try to add the attribute \*before\*
defining the body&#8203;:

$ perl \-lwe 'my $x; sub foo :lvalue; sub foo \{ $x \} $x=1; print foo; foo=123; print $x'
Can't modify non\-lvalue subroutine call in scalar assignment at \-e line 1\, near "123;"
Execution of \-e aborted due to compilation errors\.

Apparently the lvalue attribute gets lost when the body is defined
without the attribute\.  The assignment op sees that it's not an lvalue\.
And here's a related situation&#8203;:

$ perl \-lwe 'my $x; sub foo :lvalue; $x=1; print foo; foo=123; print $x; sub foo \{ $x \}'
1
1

As in the previous case\, the pre\-declared lvalueness is lost when the
body is defined\.  This time\, though\, by fiddling with the order of things
I'm able to trick the assignment op into thinking it's an lvalue sub\.
It behaves like the lvalue\-flagged non\-lvalue sub that I'm reporting as
a bug\.  Not sure how this case ought to behave\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.10.0:

Configured by Debian Project at Thu Jan  1 12:43:38 UTC 2009.

Summary of my perl5 (revision 5 version 10 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.6.26-1-686, archname=i486-linux-gnu-thread-multi
    uname='linux rebekka 2.6.26-1-686 #1 smp mon dec 15 18:15:07 utc 2008 i686 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN -Dcccdlflags=-fPIC -Darchname=i486-linux-gnu -Dprefix=/usr -Dprivlib=/usr/share/perl/5.10 -Darchlib=/usr/lib/perl/5.10 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.10.0 -Dsitearch=/usr/local/lib/perl/5.10.0 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Ud_ualarm -Uusesfio -Uusenm -DDEBUGGING=-g -Doptimize=-O2 -Duseshrplib -Dlibperl=libperl.so.5.10.0 -Dd_dosuid -des'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -g',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fno-strict-aliasing -pipe -I/usr/local/include'
    ccversion='', gccversion='4.3.2', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /usr/lib64
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=/lib/libc-2.7.so, so=so, useshrplib=true, libperl=libperl.so.5.10.0
    gnulibc_version='2.7'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -g -L/usr/local/lib'

Locally applied patches:
    


@INC for perl 5.10.0:
    /etc/perl
    /usr/local/lib/perl/5.10.0
    /usr/local/share/perl/5.10.0
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.10
    /usr/share/perl/5.10
    /usr/local/lib/site_perl
    .


Environment for perl 5.10.0:
    HOME=/home/zefram
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/zefram/pub/i686-pc-linux-gnu/bin:/home/zefram/pub/common/bin:/usr/bin:/usr/X11R6/bin:/bin:/usr/local/bin:/usr/games
    PERL_BADLANG (unset)
    SHELL=/usr/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
alecclews@gmail.com - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @rgs

Fixed by :

commit eac910c876359d864e67b412fa2d9f03b5ba8a0b
Author&#8203;: Gerard Goossen \<gerard@&#8203;ggoossen\.net>
Date&#8203;:   Sun Dec 27 18&#8203;:02&#8203;:37 2009 \+0100

&nbsp;    Retain builtin attributes from pre\-declaration\. Fixes \[perl \#68758\]\.

 op\.c            |    3 \+\+\-
 t/op/attrs\.t    |    7 \+\+\+\+\+\+\-
 t/op/sub\_lval\.t |   10 \+\+\+\+\+\+\+\+\+\-
 3 files changed\, 17 insertions\(\+\)\, 3 deletions\(\-\)

commit 885ef6f56b61fd750ef3b1fa614d11480baac635
Author&#8203;: Gerard Goossen \<gerard@&#8203;ggoossen\.net>
Date&#8203;:   Sun Dec 27 17&#8203;:24&#8203;:28 2009 \+0100

&nbsp;    Ignore a lvalue attribute after the subroutine has been defined\, and
warn about it\. Fixes part of \[perl \#68758\]\.

 op\.c             |    4 \+\+\+\-
 pod/perldiag\.pod |    8 \+\+\+\+\+\+\+\+
 t/op/attrs\.t     |    8 \+\+\+\+\-\-\-\-
 t/op/sub\_lval\.t  |   14 \+\+\+\+\+\+\+\+\+\+\+\+\+\-
 4 files changed\, 28 insertions\(\+\)\, 6 deletions\(\-\)

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @rgarcia

2009/12/27 Gerard Goossen \<gerard@&#8203;ggoossen\.net>&#8203;:
> Retain builtin attributes from pre\-declaration\. Fixes \[perl \#68758\]\.

Thanks\, applied\, as well as the previous patch :

commit eac910c876359d864e67b412fa2d9f03b5ba8a0b
Author&#8203;: Gerard Goossen \<gerard@&#8203;ggoossen\.net>
Date&#8203;:   Sun Dec 27 18&#8203;:02&#8203;:37 2009 \+0100

&nbsp;    Retain builtin attributes from pre\-declaration\. Fixes \[perl \#68758\]\.

 op\.c            |    3 \+\+\-
 t/op/attrs\.t    |    7 \+\+\+\+\+\+\-
 t/op/sub\_lval\.t |   10 \+\+\+\+\+\+\+\+\+\-
 3 files changed\, 17 insertions\(\+\)\, 3 deletions\(\-\)

commit 885ef6f56b61fd750ef3b1fa614d11480baac635
Author&#8203;: Gerard Goossen \<gerard@&#8203;ggoossen\.net>
Date&#8203;:   Sun Dec 27 17&#8203;:24&#8203;:28 2009 \+0100

&nbsp;    Ignore a lvalue attribute after the subroutine has been defined\,
and warn about it\. Fixes part of \[perl \#68758\]\.

 op\.c             |    4 \+\+\+\-
 pod/perldiag\.pod |    8 \+\+\+\+\+\+\+\+
 t/op/attrs\.t     |    8 \+\+\+\+\-\-\-\-
 t/op/sub\_lval\.t  |   14 \+\+\+\+\+\+\+\+\+\+\+\+\+\-
 4 files changed\, 28 insertions\(\+\)\, 6 deletions\(\-\)
